components/org.wso2.carbon.identity.api.user.recovery/org.wso2.carbon.identity.api.user.recovery.commons/src/main/java/org/wso2/carbon/identity/api/user/recovery/commons/UserAccountRecoveryServiceDataHolder.java

```diff
@@ -18,54 +18,60 @@
 
 package org.wso2.carbon.identity.api.user.recovery.commons;
 
+import org.wso2.carbon.context.PrivilegedCarbonContext;
 import org.wso2.carbon.identity.recovery.internal.service.impl.password.PasswordRecoveryManagerImpl;
 import org.wso2.carbon.identity.recovery.services.password.PasswordRecoveryManager;
 import org.wso2.carbon.identity.recovery.services.username.UsernameRecoveryManager;
 
 import java.util.List;
+import java.util.stream.Collectors;
 
 /**
  * Service holder class for user account recovery.
  */
 public class UserAccountRecoveryServiceDataHolder {
 
-    private static UsernameRecoveryManager usernameRecoveryManager;
-    private static List<PasswordRecoveryManager> passwordRecoveryManager;
+    private UserAccountRecoveryServiceDataHolder() {
 
-    /**
-     * Get UsernameRecoveryManager instance.
-     *
-     * @return UsernameRecoveryManager
-     */
-    public static UsernameRecoveryManager getUsernameRecoveryManager() {
+    }
+
+    private static class UsernameRecoveryManagerHolder {
+
+        static final UsernameRecoveryManager SERVICE = (UsernameRecoveryManager) PrivilegedCarbonContext
+                .getThreadLocalCarbonContext().getOSGiService(UsernameRecoveryManager.class, null);
+    }
+
+    private static class PasswordRecoveryManagerHolder {
 
-        return UserAccountRecoveryServiceDataHolder.usernameRecoveryManager;
+        static final List<PasswordRecoveryManager> SERVICES = PrivilegedCarbonContext.getThreadLocalCarbonContext()
```

Comment:
Shall we add a log when the list is empty to preserve previous behavior to a point. (As we can't throw an exception from here)